### PR TITLE
server: emit disk slowness detected/cleared events in eventlog

### LINKED ISSUE #125977: storage: log disk stalls to system.eventlog
Perusing the "events" window in the DB console is a common initial flow when diagnosing issues to make sure nothing obvious correlates with timeline of an issue. Disk slowness and stalls are not currently logged in the event view (which is sourced from `system.eventlog` table). Currently disk slowness and stalls are only logged in pebble logs.

Ask here is that the we log an event into `system.eventlog`. Log when the slowness is first detected, and then when it clears log the event:

- Timestamp
- EventType: disk_slowness_detected; disk_slowness_cleared
- NodeID: node
- StoreID: store_id

Note:  Writing to the event log in these cases should be best effort, as if there's a disk stall, and the leaseholder for the event log is on the same node, we think it might deadlock. 

Jira issue: CRDB-39694

### PR DESCRIPTION
**Pull request by itsbilal**
Previously, disk slowness events would show up in metrics as well as in the pebble log, but not as a structured event or in the system.eventlog table. This change adds two events: a DiskSlownessDetected and a DiskSlownessCleared event. The former is emitted when disk slowness is first observed on a node, and the latter is emitted when no disk slowness events have been observed on a node for more than 30 seconds.

Epic: none
Fixes: #125977

Release note (ops change): Add eventlog events for disk slowness detected and disk slowness cleared on a store.

### CODE REVIEW

**Review by RaduBerinde — APPROVED** [review]
:lgtm: 

An alternative to having Stop/Start events would be to have a single event which refers to a timeline - "DiskSlownessDetected n times from time X to time Y (max latency observed: XX)". The initial event would have a very short timeline (though perhaps we'd want to wait some seconds to see if we get more events) and then each event would have a 30s timeline or similar. We can do this later, just putting it out there (I was just discussing something along these lines with @dhartunian and @kyle-a-wong)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127025#-:-O1YTinADkptFPJcfVcl:blw3i32)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @anish-shanbhag, @itsbilal, @jbowens, and @kyle-a-wong)

___
*[`pkg/server/node.go` line 883 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127025#-O1YSRKq4YOuUEDkYZ7K:-O1YSRKq4YOuUEDkYZ7L:b-nampzq) ([raw file](https://github.com/cockroachdb/cockroach/blob/631340b23b7bd7ef323a22498e2ef926dc430ae1/pkg/server/node.go#L883)):*
> ```Go
> 	// slowness cleared event. Any other disk slow events for the same store in
> 	// between are ignored as duplicates.
> 	const diskSlowClearInterval = 30 * time.Second
> ```

We can leave a TODO here to keep some statistics about those ignored events - how many stalls, average/max duration etc
___
*[`pkg/server/node.go` line 892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127025#-O1YSou45l8fCvFZiLFJ:-O1YSou45l8fCvFZiLFK:b-buuwee) ([raw file](https://github.com/cockroachdb/cockroach/blob/631340b23b7bd7ef323a22498e2ef926dc430ae1/pkg/server/node.go#L892)):*
> ```Go
> 	n.diskSlowCoalescerMu.lastDiskSlow[storeID] = timeutil.Now()
> 	n.diskSlowCoalescerMu.Unlock()
> 	if ok && timeutil.Since(last) < diskSlowClearInterval {
> ```

I think it would be nicer if we stored the current state - whether we are currently waiting to coalesce events or not. That way it will be clear we will never run more than one AsyncTask in parallel.
___
*[`pkg/server/node.go` line 909 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127025#-O1YTLcK-CMDHnsBBZ3J:-O1YTLcLE3IIAJCnRlYy:bcilhkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/631340b23b7bd7ef323a22498e2ef926dc430ae1/pkg/server/node.go#L909)):*
> ```Go
> 		for {
> 			select {
> 			case <-ctx.Done():
> ```

This is the context of the code path which registered the slowness in Pebble, no? I don't think we should depend on that context being Done. I think we should rely on the stopper.

**Review by itsbilal — COMMENTED** [review]
TFTR!

I didn't want to deviate in the type / content of the event from what's described in #125977. I believe there are external reasons for it being an explicit start/stop. Plus there's value in emitting the start event ASAP instead of waiting to coalesce some more events, by when we could have crashed.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127025#-:-O1YZp9pDGPWUsaXB_L7:b-pw2qo9)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @anish-shanbhag, @jbowens, @kyle-a-wong, and @RaduBerinde)

___
*[`pkg/server/node.go` line 883 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127025#-O1YSRKq4YOuUEDkYZ7K:-O1YZWov9Q8LLGRAf2lR:b-wxj54a) ([raw file](https://github.com/cockroachdb/cockroach/blob/631340b23b7bd7ef323a22498e2ef926dc430ae1/pkg/server/node.go#L883)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We can leave a TODO here to keep some statistics about those ignored events - how many stalls, average/max duration etc
</blockquote></details>

The number of slowness events are already tracked in `storage.disk-slow`, but yeah the durations of the stalls aren't really tracked anywhere. Added.
___
*[`pkg/server/node.go` line 892 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127025#-O1YSou45l8fCvFZiLFJ:-O1YaXYi5meugelDTYQF:bf7c8cs) ([raw file](https://github.com/cockroachdb/cockroach/blob/631340b23b7bd7ef323a22498e2ef926dc430ae1/pkg/server/node.go#L892)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think it would be nicer if we stored the current state - whether we are currently waiting to coalesce events or not. That way it will be clear we will never run more than one AsyncTask in parallel.
</blockquote></details>

Good point. I am now using the zero value of time.Time to signify that there's no async task running.
___
*[`pkg/server/node.go` line 909 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127025#-O1YTLcK-CMDHnsBBZ3J:-O1YZo6Y1aYr-iQFvoFC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/631340b23b7bd7ef323a22498e2ef926dc430ae1/pkg/server/node.go#L909)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This is the context of the code path which registered the slowness in Pebble, no? I don't think we should depend on that context being Done. I think we should rely on the stopper.
</blockquote></details>

Done.

**Review by RaduBerinde — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127025#-:-O1b8r6WB3ldLiW4hzkz:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @anish-shanbhag, @itsbilal, @jbowens, and @kyle-a-wong)

___
*[`pkg/server/node.go` line 920 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/127025#-O1b8Lpg3H-p7nLAHr0N:-O1b8Lpg3H-p7nLAHr0O:b-rgeli7) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e2a10b5f1bdcc3b74bd06259d3935085613a9c/pkg/server/node.go#L920)):*
> ```Go
> 				if timeutil.Since(last) < diskSlowClearInterval {
> 					clearing = true
> 					n.diskSlowCoalescerMu.lastDiskSlow[storeID] = time.Time{}
> ```

We could just delete from the map here and then we don't have to check for zero Time.
___
*[`pkg/ui/workspaces/db-console/src/util/eventTypes.ts` line 142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/127025#-O1b8WC39YLg1lLu2aVL:-O1b8WC39YLg1lLu2aVM:b-yjqxhp) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e2a10b5f1bdcc3b74bd06259d3935085613a9c/pkg/ui/workspaces/db-console/src/util/eventTypes.ts#L142)):*
> ```XML
> // Recorded when a disk slowness event is detected on a store.
> export const DISK_SLOWNESS_DETECTED = "disk_slowness_detected";
> // Recorded when a disk slowness event has not been detected.
> ```

[nit] "has not been detected" doesn't sound right.. Maybe "Recorded when disk slowness is no longer happening after a previously reported slowness detected event"

**Review by itsbilal — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127025#-:-O1baEZF-BaMyllLRiuv:b-814g8k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @anish-shanbhag, @jbowens, and @kyle-a-wong)

___
*[`pkg/server/node.go` line 920 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/127025#-O1b8Lpg3H-p7nLAHr0N:-O1baFR9EOI7qFsX2SNo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e2a10b5f1bdcc3b74bd06259d3935085613a9c/pkg/server/node.go#L920)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We could just delete from the map here and then we don't have to check for zero Time.
</blockquote></details>

Done.
___
*[`pkg/ui/workspaces/db-console/src/util/eventTypes.ts` line 142 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/127025#-O1b8WC39YLg1lLu2aVL:-O1baGWQ55QChTEkpI9X:b-376vyb) ([raw file](https://github.com/cockroachdb/cockroach/blob/19e2a10b5f1bdcc3b74bd06259d3935085613a9c/pkg/ui/workspaces/db-console/src/util/eventTypes.ts#L142)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] "has not been detected" doesn't sound right.. Maybe "Recorded when disk slowness is no longer happening after a previously reported slowness detected event"
</blockquote></details>

Done. Good catch.

File: pkg/server/node.go
Comment (nicktrav) [inline pkg/server/node.go:920]: Don't we want to clear if the time since the last disk slow event on the store is _greater_ than the clear interval? i.e. it's been > 30s since the last event, therefore we're clear.

File: pkg/server/node.go
Comment (nicktrav) [inline pkg/server/node.go:887]: Do we want to make this configurable? I imagine there will be differing interpretations of what it means for slowness to clear.

File: pkg/server/node.go
Comment (nicktrav) [inline pkg/server/node.go:936]: Is this blocking, or best effort? What happens if we're attempting to log slowness for the store that is the LH for the event log, and its disk is indefinitely stalled?

Comment (itsbilal) [reply pkg/server/node.go:936]: This method does a log to the node's logger (where a stall would have a different flushing goroutine wait instead), and it does an INSERT asynchronously in a different goroutine into the eventlog table. So it seems like we should be covered here, and shouldn't unintentionally hold up the disk health checker.

File: pkg/server/node.go
Comment (itsbilal) [reply pkg/server/node.go:887]: We already have maxSyncDuration that's configurable, and we can make this a multiple of that (or just equal to that). What do you think about that idea?

File: pkg/server/node.go
Comment (itsbilal) [reply pkg/server/node.go:920]: Good catch! This is an artifact from a previous refactor 🙈 Fixed.

File: pkg/server/node.go
Comment (nicktrav) [reply pkg/server/node.go:887]: > or just equal to that

This seems reasonable. If the stall is longer than this, we'll crash, right?

**Review by nicktrav — APPROVED** [review]
LGTM modulo the last couple comments.

File: pkg/server/node.go
Comment (itsbilal) [reply pkg/server/node.go:936]: On closer inspection the `log.StructuredEvent` could under some cases write directly to the logger file sink, even though the SQL write will always be async. So to decouple this from the handler coming all the way from Pebble, I moved the entire `logStructuredEvent` call into the async task below, and left a comment about it.

File: pkg/server/node.go
Comment (itsbilal) [reply pkg/server/node.go:887]: Not necessarily. If we're seeing write events that are consistently slow at about 5s or so per request, we'll see a prolonged disk slowness event in the event log, but no crash because no individual write took >20s.

But you're right, that more commonly we either see slowness events around 5s that go away quickly, or >20s in which case there's just a crash. So I made the `maxSyncDuration` change.

**Comment by itsbilal** [comment]
Given the last two LGTMs, I think this is ready to merge.

bors r=nicktrav

**Pull request merged**
